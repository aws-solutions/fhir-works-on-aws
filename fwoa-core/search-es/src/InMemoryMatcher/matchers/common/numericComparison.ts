/*
 *  Copyright Amazon.com, Inc. or its affiliates. All Rights Reserved.
 *  SPDX-License-Identifier: Apache-2.0
 *
 */

export interface Range {
  start: number;
  end: number;
}

export const compareNumberToRange = (
  prefix: string,
  searchParamRange: Range,
  resourceValue: number
): boolean => {
  const { start, end } = searchParamRange;
  // See https://www.hl7.org/fhir/search.html#prefix
  switch (prefix) {
    case 'ne': // not equal
      return resourceValue < start || resourceValue > end;
    case 'eq': // equal
      return resourceValue >= start && resourceValue <= end;
    case 'lt': // less than
      return resourceValue < end;
    case 'le': // less or equal
      return resourceValue <= end;
    case 'gt': // greater than
      return resourceValue > start;
    case 'ge': // greater or equal
      return resourceValue >= start;
    case 'sa': // starts after
      return resourceValue > end;
    case 'eb': // ends before
      return resourceValue < start;
    case 'ap': // approximately
      // same as eq for now
      return resourceValue >= start && resourceValue <= end;
    default:
      // this should never happen
      throw new Error(`unknown search prefix: ${prefix}`);
  }
};

// See the interpretation of prefixes when applied to ranges: https://www.hl7.org/fhir/search.html#prefix. It can be counterintuitive at first
export const compareRanges = (prefix: string, searchParamRange: Range, resourceRange: Range): any => {
  const { start, end } = searchParamRange;
  const resourceStart = resourceRange.start;
  const resourceEnd = resourceRange.end;
  // See https://www.hl7.org/fhir/search.html#prefix
  switch (prefix) {
    case 'ne': // not equal
      return !(resourceStart >= start && resourceEnd <= end);
    case 'eq': // equal
      return resourceStart >= start && resourceEnd <= end;
    case 'lt': // less than
    case 'le': // less or equal
      return resourceStart <= end;
    case 'gt': // greater than
    case 'ge': // greater or equal
      return resourceEnd >= start;
    case 'sa': // starts after
      return resourceStart > end;
    case 'eb': // ends before
      return resourceEnd < start;
    case 'ap': // approximately
      return !(resourceStart > end) && !(resourceEnd < start);
    default:
      // this should never happen
      throw new Error(`unknown search prefix: ${prefix}`);
  }
};

/**
 * When a comparison prefix in the set lgt, lt, ge, le, sa & eb is provided, the implicit precision of the number is ignored,
 * and they are treated as if they have arbitrarily high precision. https://www.hl7.org/fhir/search.html#number
 * @param prefix
 * @param number
 * @param range
 */
export const applyPrefixRulesToRange = (prefix: string, number: number, range: Range): Range => {
  if (prefix === 'eq' || prefix === 'ne') {
    return range;
  }

  return { start: number, end: number };
};
