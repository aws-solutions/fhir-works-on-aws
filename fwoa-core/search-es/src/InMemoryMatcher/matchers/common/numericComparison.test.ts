/*
 *  Copyright Amazon.com, Inc. or its affiliates. All Rights Reserved.
 *  SPDX-License-Identifier: Apache-2.0
 *
 */

import { compareNumberToRange, compareRanges } from './numericComparison';

const prefixes = ['ne', 'eq', 'lt', 'le', 'gt', 'ge', 'sa', 'eb', 'ap'];

describe('compareNumberToRange', () => {
  const resourceValue = 10;
  const ranges = [
    [8, 9],
    [9, 10],
    [9, 11],
    [10, 11],
    [11, 12]
  ];
  test.each(prefixes)('%p', (prefix) => {
    const results = ranges.map(
      ([start, end]) =>
        `(${resourceValue}) ${prefix} (${start}, ${end}) is ${compareNumberToRange(
          prefix,
          { start, end },
          resourceValue
        )}`
    );

    expect(results).toMatchSnapshot();
  });
});

describe('compareRanges', () => {
  const resourceRange = {
    lowerBound: 10,
    upperBound: 20
  };

  const ranges = [
    [8, 9],
    [9, 10],
    [9, 15],
    [9, 20],
    [9, 21],
    [10, 15],
    [10, 20],
    [10, 21],
    [11, 19],
    [15, 20],
    [15, 21],
    [20, 21],
    [21, 22]
  ];

  console.log(ranges);

  test.each(prefixes)('%p', (prefix) => {
    const results = ranges.map(
      ([start, end]) =>
        `(${resourceRange.lowerBound}, ${
          resourceRange.upperBound
        }) ${prefix} (${start}, ${end}) is ${compareRanges(
          prefix,
          { start, end },
          { start: resourceRange.lowerBound, end: resourceRange.upperBound }
        )}`
    );

    expect(results).toMatchSnapshot();
  });
});
