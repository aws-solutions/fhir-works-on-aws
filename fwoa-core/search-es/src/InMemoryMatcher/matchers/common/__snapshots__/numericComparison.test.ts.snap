// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`compareNumberToRange "ap" 1`] = `
Array [
  "(10) ap (8, 9) is false",
  "(10) ap (9, 10) is true",
  "(10) ap (9, 11) is true",
  "(10) ap (10, 11) is true",
  "(10) ap (11, 12) is false",
]
`;

exports[`compareNumberToRange "eb" 1`] = `
Array [
  "(10) eb (8, 9) is false",
  "(10) eb (9, 10) is false",
  "(10) eb (9, 11) is false",
  "(10) eb (10, 11) is false",
  "(10) eb (11, 12) is true",
]
`;

exports[`compareNumberToRange "eq" 1`] = `
Array [
  "(10) eq (8, 9) is false",
  "(10) eq (9, 10) is true",
  "(10) eq (9, 11) is true",
  "(10) eq (10, 11) is true",
  "(10) eq (11, 12) is false",
]
`;

exports[`compareNumberToRange "ge" 1`] = `
Array [
  "(10) ge (8, 9) is true",
  "(10) ge (9, 10) is true",
  "(10) ge (9, 11) is true",
  "(10) ge (10, 11) is true",
  "(10) ge (11, 12) is false",
]
`;

exports[`compareNumberToRange "gt" 1`] = `
Array [
  "(10) gt (8, 9) is true",
  "(10) gt (9, 10) is true",
  "(10) gt (9, 11) is true",
  "(10) gt (10, 11) is false",
  "(10) gt (11, 12) is false",
]
`;

exports[`compareNumberToRange "le" 1`] = `
Array [
  "(10) le (8, 9) is false",
  "(10) le (9, 10) is true",
  "(10) le (9, 11) is true",
  "(10) le (10, 11) is true",
  "(10) le (11, 12) is true",
]
`;

exports[`compareNumberToRange "lt" 1`] = `
Array [
  "(10) lt (8, 9) is false",
  "(10) lt (9, 10) is false",
  "(10) lt (9, 11) is true",
  "(10) lt (10, 11) is true",
  "(10) lt (11, 12) is true",
]
`;

exports[`compareNumberToRange "ne" 1`] = `
Array [
  "(10) ne (8, 9) is true",
  "(10) ne (9, 10) is false",
  "(10) ne (9, 11) is false",
  "(10) ne (10, 11) is false",
  "(10) ne (11, 12) is true",
]
`;

exports[`compareNumberToRange "sa" 1`] = `
Array [
  "(10) sa (8, 9) is true",
  "(10) sa (9, 10) is false",
  "(10) sa (9, 11) is false",
  "(10) sa (10, 11) is false",
  "(10) sa (11, 12) is false",
]
`;

exports[`compareRanges "ap" 1`] = `
Array [
  "(10, 20) ap (8, 9) is false",
  "(10, 20) ap (9, 10) is true",
  "(10, 20) ap (9, 15) is true",
  "(10, 20) ap (9, 20) is true",
  "(10, 20) ap (9, 21) is true",
  "(10, 20) ap (10, 15) is true",
  "(10, 20) ap (10, 20) is true",
  "(10, 20) ap (10, 21) is true",
  "(10, 20) ap (11, 19) is true",
  "(10, 20) ap (15, 20) is true",
  "(10, 20) ap (15, 21) is true",
  "(10, 20) ap (20, 21) is true",
  "(10, 20) ap (21, 22) is false",
]
`;

exports[`compareRanges "eb" 1`] = `
Array [
  "(10, 20) eb (8, 9) is false",
  "(10, 20) eb (9, 10) is false",
  "(10, 20) eb (9, 15) is false",
  "(10, 20) eb (9, 20) is false",
  "(10, 20) eb (9, 21) is false",
  "(10, 20) eb (10, 15) is false",
  "(10, 20) eb (10, 20) is false",
  "(10, 20) eb (10, 21) is false",
  "(10, 20) eb (11, 19) is false",
  "(10, 20) eb (15, 20) is false",
  "(10, 20) eb (15, 21) is false",
  "(10, 20) eb (20, 21) is false",
  "(10, 20) eb (21, 22) is true",
]
`;

exports[`compareRanges "eq" 1`] = `
Array [
  "(10, 20) eq (8, 9) is false",
  "(10, 20) eq (9, 10) is false",
  "(10, 20) eq (9, 15) is false",
  "(10, 20) eq (9, 20) is true",
  "(10, 20) eq (9, 21) is true",
  "(10, 20) eq (10, 15) is false",
  "(10, 20) eq (10, 20) is true",
  "(10, 20) eq (10, 21) is true",
  "(10, 20) eq (11, 19) is false",
  "(10, 20) eq (15, 20) is false",
  "(10, 20) eq (15, 21) is false",
  "(10, 20) eq (20, 21) is false",
  "(10, 20) eq (21, 22) is false",
]
`;

exports[`compareRanges "ge" 1`] = `
Array [
  "(10, 20) ge (8, 9) is true",
  "(10, 20) ge (9, 10) is true",
  "(10, 20) ge (9, 15) is true",
  "(10, 20) ge (9, 20) is true",
  "(10, 20) ge (9, 21) is true",
  "(10, 20) ge (10, 15) is true",
  "(10, 20) ge (10, 20) is true",
  "(10, 20) ge (10, 21) is true",
  "(10, 20) ge (11, 19) is true",
  "(10, 20) ge (15, 20) is true",
  "(10, 20) ge (15, 21) is true",
  "(10, 20) ge (20, 21) is true",
  "(10, 20) ge (21, 22) is false",
]
`;

exports[`compareRanges "gt" 1`] = `
Array [
  "(10, 20) gt (8, 9) is true",
  "(10, 20) gt (9, 10) is true",
  "(10, 20) gt (9, 15) is true",
  "(10, 20) gt (9, 20) is true",
  "(10, 20) gt (9, 21) is true",
  "(10, 20) gt (10, 15) is true",
  "(10, 20) gt (10, 20) is true",
  "(10, 20) gt (10, 21) is true",
  "(10, 20) gt (11, 19) is true",
  "(10, 20) gt (15, 20) is true",
  "(10, 20) gt (15, 21) is true",
  "(10, 20) gt (20, 21) is true",
  "(10, 20) gt (21, 22) is false",
]
`;

exports[`compareRanges "le" 1`] = `
Array [
  "(10, 20) le (8, 9) is false",
  "(10, 20) le (9, 10) is true",
  "(10, 20) le (9, 15) is true",
  "(10, 20) le (9, 20) is true",
  "(10, 20) le (9, 21) is true",
  "(10, 20) le (10, 15) is true",
  "(10, 20) le (10, 20) is true",
  "(10, 20) le (10, 21) is true",
  "(10, 20) le (11, 19) is true",
  "(10, 20) le (15, 20) is true",
  "(10, 20) le (15, 21) is true",
  "(10, 20) le (20, 21) is true",
  "(10, 20) le (21, 22) is true",
]
`;

exports[`compareRanges "lt" 1`] = `
Array [
  "(10, 20) lt (8, 9) is false",
  "(10, 20) lt (9, 10) is true",
  "(10, 20) lt (9, 15) is true",
  "(10, 20) lt (9, 20) is true",
  "(10, 20) lt (9, 21) is true",
  "(10, 20) lt (10, 15) is true",
  "(10, 20) lt (10, 20) is true",
  "(10, 20) lt (10, 21) is true",
  "(10, 20) lt (11, 19) is true",
  "(10, 20) lt (15, 20) is true",
  "(10, 20) lt (15, 21) is true",
  "(10, 20) lt (20, 21) is true",
  "(10, 20) lt (21, 22) is true",
]
`;

exports[`compareRanges "ne" 1`] = `
Array [
  "(10, 20) ne (8, 9) is true",
  "(10, 20) ne (9, 10) is true",
  "(10, 20) ne (9, 15) is true",
  "(10, 20) ne (9, 20) is false",
  "(10, 20) ne (9, 21) is false",
  "(10, 20) ne (10, 15) is true",
  "(10, 20) ne (10, 20) is false",
  "(10, 20) ne (10, 21) is false",
  "(10, 20) ne (11, 19) is true",
  "(10, 20) ne (15, 20) is true",
  "(10, 20) ne (15, 21) is true",
  "(10, 20) ne (20, 21) is true",
  "(10, 20) ne (21, 22) is true",
]
`;

exports[`compareRanges "sa" 1`] = `
Array [
  "(10, 20) sa (8, 9) is true",
  "(10, 20) sa (9, 10) is false",
  "(10, 20) sa (9, 15) is false",
  "(10, 20) sa (9, 20) is false",
  "(10, 20) sa (9, 21) is false",
  "(10, 20) sa (10, 15) is false",
  "(10, 20) sa (10, 20) is false",
  "(10, 20) sa (10, 21) is false",
  "(10, 20) sa (11, 19) is false",
  "(10, 20) sa (15, 20) is false",
  "(10, 20) sa (15, 21) is false",
  "(10, 20) sa (20, 21) is false",
  "(10, 20) sa (21, 22) is false",
]
`;
